Sources/TensorFlow/Layers/Recurrent.swift

```diff
@@ -444,10 +451,61 @@ public struct RecurrentLayer<Cell: RecurrentLayerCell>: Layer {
   }
 }
 
+public protocol Mergable 
+where Self: Differentiable {
+  @differentiable
+  static func +(lhs: Self, rhs: Self) -> Self
+}
+
+public enum MergeMode {
+  case sum
+}
+
+public struct BidirectionalRecurrentLayer<Cell: RecurrentLayerCell>: Layer 
+where Cell.TimeStepOutput: Mergable {
+  public typealias Input = [Cell.TimeStepInput]
+  public typealias Output = [Cell.TimeStepOutput]
+  public typealias Merge = @differentiable (Cell.TimeStepOutput, Cell.TimeStepOutput) -> Cell.TimeStepOutput
+
+  @noDerivative public let mergeMode: MergeMode
+  public var forward, backward: RecurrentLayer<Cell>
+
+  public init(_ cell: @autoclosure () -> Cell, mergeMode: MergeMode = .sum) {
```

Comment:
Rather than using a `MergeMode` enum, how about directly taking merge mode as a `@differentiable` closure instead? Closures are more flexible than enums, which are not extensible.
```suggestion
  public init(_ cell: @autoclosure () -> Cell, mergeMode: @differentiable (Cell.TimeStepOutput, Cell.TimeStepOutput) -> Cell.TimeStepOutput = sum) {
```
This should also allow you to remove `protocol Mergable`.

Reply:
Yes, I wanted to, but I hit the issue described above, and I wanted to do a PR with working tests. I will try it again, but if I will not be able to fix it, can I do a PR with a runtime error and perhaps you will be able to correct me?

Reply:
Ah, I see. Yes, I can help take a look if you share a branch reproducing the runtime error.

Reply:
I updated PR. I think the issue is that Swift can not take right `sum` function? For example if I put default function to the init: `merge: @escaping MergeFunction = sum` it crashes on 

```swift
/Users/peter.jung/Projects/swift-apis/Sources/TensorFlow/Layers/Recurrent.swift:484:81: error: no exact matches in reference to global function 'sum'
  public init(_ cell: @autoclosure () -> Cell, merge: @escaping MergeFunction = sum) {
                                                                                ^
/Users/peter.jung/Projects/swift-apis/Sources/TensorFlow/Layers/Recurrent.swift:450:13: note: found this candidate
public func sum<Scalar: TensorFlowFloatingPoint>(
            ^
/Users/peter.jung/Projects/swift-apis/Sources/TensorFlow/Layers/Recurrent.swift:458:13: note: found this candidate
public func sum<Scalar: TensorFlowFloatingPoint>(
            ^
/Users/peter.jung/Projects/swift-apis/Sources/TensorFlow/Layers/Recurrent.swift:469:13: note: found this candidate
public func sum<Scalar: TensorFlowFloatingPoint>(
            ^
[1/3] Compiling TensorFlow Recurrent.swift
```

Reply:
Ah, I see. You've added three concrete overloads of `sum` for various `RecurrentLayerCell.TimeStepOutput` types:

<details>
<p>

```swift
@differentiable
public func sum<Scalar: TensorFlowFloatingPoint>(
  _ firstOutput: BasicRNNCell<Scalar>.TimeStepOutput, 
  _ secondOutput: BasicRNNCell<Scalar>.TimeStepOutput
) -> BasicRNNCell<Scalar>.TimeStepOutput {
  return BasicRNNCell<Scalar>.TimeStepOutput(firstOutput.value + secondOutput.value)
}

@differentiable
public func sum<Scalar: TensorFlowFloatingPoint>(
  _ firstOutput: LSTMCell<Scalar>.TimeStepOutput,
  _ secondOutput: LSTMCell<Scalar>.TimeStepOutput
) -> LSTMCell<Scalar>.TimeStepOutput {
  return LSTMCell<Scalar>.TimeStepOutput(
    cell: firstOutput.cell + secondOutput.cell, 
    hidden: firstOutput.hidden + secondOutput.hidden
  )
}

@differentiable
public func sum<Scalar: TensorFlowFloatingPoint>(
  _ firstOutput: GRUCell<Scalar>.TimeStepOutput,
  _ secondOutput: GRUCell<Scalar>.TimeStepOutput
) -> GRUCell<Scalar>.TimeStepOutput {
  return GRUCell<Scalar>.TimeStepOutput(hidden: firstOutput.hidden + secondOutput.hidden)
}
```

</p>
</details>

However, you need a generic "sum" function in order pass type-checking - `RecurrentLayerCell.TimeStepOutput` must be constrained to a protocol that defines a "sum" function.

Previously, you had a `protocol Mergable` representing this. However, looking at the implementations of `TimeStepOutput.sum`, the function is really just performing memberwise  `AdditiveArithmetic.+`.

---

Can you try adding a constraint:
```swift
public struct BidirectionalRecurrentLayer<Cell: RecurrentLayerCell>: Layer
where Cell.TimeStepOutput: AdditiveArithmetic {
```

And using `+` as the default parameter value for `mergeMode`:
```swift
  public init(_ cell: @autoclosure () -> Cell, mergeMode: @differentiable (Cell.TimeStepOutput, Cell.TimeStepOutput) -> Cell.TimeStepOutput = +) {
```

Reply:
If we want to extend "merge mode" to support other functions like concatenation, which aren't defined as requirements of any pre-existing protocol, we will need to add a new protocol defining all "merge functions" as requirements.

To make incremental progress, we can look into extending "merge mode" in a follow-up PR! Using and testing just `+` for now is totally reasonable.